File: cpp/ecbenchmark/src/operator/Shifted.cpp
#include "ecbenchmark/operator/Shifted.h"

namespace ecb {

    Shifted::Shifted(const std::vector<scalar>& shift, int fromIndex,  Function* f)
    : Operator(f), _shift(shift), _fromIndex(fromIndex){

    }
    

    void Shifted::setShift(const std::vector<scalar>& shift) {
        this->_shift = shift;
    }

    std::vector<scalar> Shifted::shift() const {
        return this->_shift;
    }

    void Shifted::setFromIndex(int index) {
        this->_fromIndex = index;
    }

    int Shifted::fromIndex() const {
        return this->_fromIndex;
    }

    scalar Shifted::f(const std::vector<scalar>& x) {
        std::vector<scalar> shifted_x;
        shifted_x.resize(x.size());
        int shiftIndex = _fromIndex;
        for (size_t i = 0; i < x.size(); ++i) {
            shifted_x[i] = x[i] - _shift[shiftIndex++];
        }
        return _innerFunction->f(shifted_x);
    }
}

File: java/ecbenchmark/src/ecbenchmark/operator/Shifted.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package ecbenchmark.operator;

import ecbenchmark.Operator;
import ecbenchmark.Function;

/**
 *
 * @author jcrada
 */
public class Shifted extends Operator {

    private double[] shift;
    private int fromIndex;

    public Shifted(double[] shift, int fromIndex, Function f) {
        super(f);
        this.shift = shift;
        this.fromIndex = fromIndex;
    }

    @Override
    public double f(final double[] x) {
        double[] shifted_x = new double[x.length];
        int shiftIndex = fromIndex;
        for (int i = 0; i < x.length; ++i) {
            shifted_x[i] = x[i] - shift[shiftIndex++];
        }
        return innerFunction.f(shifted_x);
    }
}
